streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -354,7 +359,7 @@ public void close() {
 
     @Override
     public Map<TopicPartition, Long> offsets() {
-        return Collections.unmodifiableMap(offsets);
+        return Collections.unmodifiableMap(new HashMap<>(offsets));
```

Comment:
Making a deep copy and making it read-only seems to be redundant to me -- why do we need both?

Reply:
Consider the case where you get `offsets` and then read from it once, and then you tried to do sth with it later but since it is a view and its values have been changed without the caller noticing, given its call frequency I think it is safer to make the returned values final.